tool/src/org/antlr/v4/codegen/CodeGenerator.java

```diff
@@ -37,70 +36,77 @@
 
 	public final Tool tool;
 
-	public final String language;
+	public final TargetType targetType;
 
 	private Target target;
 
 	public int lineWidth = 72;
 
-	private CodeGenerator(String language) {
+	public CodeGenerator(TargetType targetType) {
 		this.g = null;
 		this.tool = null;
-		this.language = language;
+		this.targetType = targetType;
 	}
 
 	public CodeGenerator(Grammar g) {
-		this(g.tool, g, g.getOptionString("language"));
+		this(g.tool, g, g.getTargetType());
 	}
 
-	public CodeGenerator(Tool tool, Grammar g, String language) {
+	public CodeGenerator(Tool tool, Grammar g, TargetType targetType) {
 		this.g = g;
 		this.tool = tool;
-		this.language = language != null ? language : DEFAULT_LANGUAGE;
+		this.targetType = targetType;
 	}
 
-	public static boolean targetExists(String language) {
-		String targetName = "org.antlr.v4.codegen.target."+language+"Target";
-		try {
-			Class<? extends Target> c = Class.forName(targetName).asSubclass(Target.class);
-			Constructor<? extends Target> ctor = c.getConstructor(CodeGenerator.class);
-			CodeGenerator gen = new CodeGenerator(language);
-			Target target = ctor.newInstance(gen);
-			return target.templatesExist();
-		}
-		catch (Exception e) { // ignore errors; we're detecting presence only
+	public Target getTarget() {
+		if (target != null || targetType == null) {
+			return target;
 		}
-		return false;
-	}
 
-
-	public Target getTarget() {
-		if ( target == null && targetExists(language) ) {
-			loadLanguageTarget(language);
+		switch (targetType) {
```

Comment:
Hmm... Why is reflection bad in this case? If we happen to create a new target we have to modify this code.

Reply:
It's better to get rid of refection in static languages like Java if it's possible because it breaks type guarantees. Yes, we have to add a couple of lines if we add a new target, but it does not look like a big problem.

Reply:
Ok, I'll revert reflection back together with `TargetType` -> `String`, because unofficial targets have its own unknown values that can be accessed only by using reflection.

Reply:
I think you are focusing on a minor code cleanliness issue, but there are bigger principles at stake here. The biggest issue is now it is less flexible and users cannot create their own custom target object.  also you've created a situation where  we have increased the number of places to change when we add a new target.  I will say that it looks like your code suggests that different targets should handle string encoding differently, which the original code didn't do.  Either we do need to fix this or this spuriously introduced flexibility we don't need.

There are two other important considerations when you are considering changes to code:

1.   You have a limited budget of my time and you spend some of it every time I have to review something; not sure this is worth spending treasure on.

2. The code change has no effect to the user and yet you are risking breaking something by changing the code for some small principle of static types that, as I said, also reduces flexibility.  Unless something is broken or you need to add a new feature, I just can't see any reason to alter code. Remember we have hundreds of thousands of users around the world; who knows what we will break.

I'd like to revert this change from a string to an enumerated type. thanks!

Currently we don't have a stable build system so I think we should focus on that.  Also, I can't possibly listen to all of the issues and PRs that command so it would be great if you could triage those for my attention. I think this was your original suggestion.

Another interesting area you want a bigger task is to think about measuring performance across targets. One simple way might be to have target templates for capturing the current time in nanoseconds or whatever. Then a performance variant of the test rig could inject that into the tests as they are generated.  We'd also need a way to run each test multiple times from within the same process rather than rebuilding and rerunning.

Reply:
> The code change has no effect to the user and yet you are risking breaking something by changing the code for some small principle of static types that, as I said, also reduces flexibility.

Originally I was fixing bugs https://github.com/antlr/antlr4/issues/2281 and https://github.com/antlr/antlr4/issues/2885. During fixing, I noticed some inconsistency (JavaScript uses the call to Java method) and redundancy. That's why I decided to refactor it together with fixing.

> Remember we have hundreds of thousands of users around the world; who knows what we will break.

I understand it quite well because I'm working on Kotlin and there are much more users. We have a lot of tests (thousands) and anyway some code is not covered, bugs sometimes appear. Maybe it makes sense to think about alpha or beta version publishing. Bugs appears always during product development.